File: LEETCODE_18_4SUM.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    vector<vector<int>> solve(int& first, int& target, vector<int>& nums) {
        set<vector<int>> result;
        int len = nums.size();
        for(int next = first+1; next < len-2; next++) {
            int start = next+1, end = len-1;
            while(start < end) {
                if(nums[first]*1LL + nums[next]*1LL + nums[start]*1LL + nums[end]*1LL == target*1LL) {
                    result.insert({nums[first], nums[next], nums[start], nums[end]});
                    start++;
                    end--;
                }
                else if(nums[first]*1LL + nums[next]*1LL + nums[start]*1LL + nums[end]*1LL < target*1LL) {
                    start++;
                }
                else {
                    end--;
                }
            }
        }
        return vector<vector<int>>(result.begin(), result.end());
    }
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        set<vector<int>> answer;
        int len = nums.size();
        for(int i = 0 ; i < len-3; i++) {
            vector<vector<int>> result = this->solve(i, target, nums);
            for(vector<int>& temp: result) {
                answer.insert(temp);
            }
        }
        return vector<vector<int>>(answer.begin(), answer.end());
    }
};

File: LEETCODE_18_4SUM.java
import java.util.*;

public class LEETCODE_18_4SUM {
    
    private List<List<Integer>> solve(int first, int target, int[] nums) {
        Set<List<Integer>> result = new HashSet<>();
        int len = nums.length;
        for(int next = first + 1; next < len - 2; next++) {
            int start = next + 1, end = len - 1;
            while (start <= end) {
                if((long)nums[first] + (long)nums[next] + (long)nums[start] + (long)nums[end] == (long)target) {
                    result.add(Arrays.asList(nums[first], nums[next], nums[start], nums[end]));
                    start++;
                    end--;
                }
                else if((long)nums[first] + (long)nums[next] + (long)nums[start] + (long)nums[end] < (long)target) {
                    start++;
                }
                else {
                    end--;
                }
            }
        }
        return new ArrayList<>(result);
    }

    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        Set<List<Integer>> answer = new HashSet<>();
        int len = nums.length;
        for(int i = 0; i < len - 3; i++) {
            List<List<Integer>> result = this.solve(i, target, nums);
            answer.addAll(result);
        }
        return new ArrayList<>(answer);
    }
}